Split out module MergingRun

mheinzel opened the pull request:
It is one of the core abstractions of the code base. This will allow us to use MergingRun in other modules without having to depend on the whole `MergeSchedule`.

I also untangled its usage a bit to provide a clean module interface. For some of the later changes, there are many alternative solutions, so I added them as separate commits that can be better viewed and discussed individually.

jorisdral commented on src/Database/LSMTree/Internal/MergingRun.hs:
It would be nice for the review of commit 8e26635abb1ddc6fb00d244160a4f218cda333aa if is split in two: one that only does the moving part, and one that makes the minor changes. Would you be willing to make that change before I review the PR?

mheinzel replied:
Done. I even split the first commit a bit further.

jorisdral reviewed (commented):
Nice refactoring! Some suggestions

dcoutts commented on src/Database/LSMTree/Internal/MergeSchedule.hs:
```diff
@@ -723,27 +722,21 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
-        let numInputRuns = NumRuns $ V.length rs
-        let numInputEntries = V.foldMap' Run.size rs
+          TraceNewMerge (V.map Run.size rs) (runNumber runPaths) caching alloc mergePolicy mergeLevel
+        -- TODO: creating the merge should happen in MR.new
```
:+1: I agree.

dcoutts commented on src/Database/LSMTree/Internal/MergeSchedule.hs:
```diff
@@ -723,27 +722,21 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
-            case mergeMaybe of
-              Nothing -> error "newMerge: merges can not be empty"
-              Just m  -> Merging <$!> MR.new mergePolicy rs m
+            let !required = MR.Credits (unNumEntries (V.foldMap' Run.size rs))
+            let !thresh = creditThresholdForLevel conf ln
+            MR.supplyCredits required thresh mr
```
Joris makes a fair point. And it's not something we'd notice easily since it'd in practice do 99.9% of the work in one shot. And maybe that's ok? Or do we need to force spending of the credits and assert it's complete?

mheinzel replied:
Yes, I remember contemplating that at some point, and yes, I do think it's the right thing to really complete the merge immediately instead of only 99.9%.

I think it works correctly here. We had a related discussion when `supplyMergeCredits` was added, which lead to adding the "We can finish the merge immediately" branch. I think that should take care of it.

But since it's not really obvious, I should at least add an assert to be sure.

mheinzel replied:
It indeed passes the assertion, but fails if I supply `required - 1` credits. Interestingly, I get a RefCount assertion failure, which is probably because the failing `isComplete` assertion leads to rolling back the temp registry, triggering the bug we discussed in the other comment. :smile:

mheinzel replied:
Instead of the slightly awkward assertion on the MergingRun's internal state, I decided to make `expectCompleted` not release the merging run, so we can use it here to 1. make sure the merge is completed and 2. access the resulting run's size for tracing.

https://github.com/IntersectMBO/lsm-tree/pull/498/commits/90cac301f287c6b39ffdf369bb205eec9cc1114c

dcoutts commented on src/Database/LSMTree/Internal/MergeSchedule.hs:
```diff
@@ -723,27 +722,21 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
+          Nothing -> error "newMerge: merges can not be empty"
+          Just m  -> allocateTemp reg (MR.new mergePolicy rs m) releaseRef
         case confMergeSchedule of
+          Incremental -> pure ()
```
Nice. Better factored, less duplication. Make it, and then either do nothing or supply enough credits now to complete in one shot. Nice.

dcoutts commented on src/Database/LSMTree/Internal/MergingRun.hs:
```diff
@@ -6,7 +6,7 @@
     MergingRun (..)
   , new
-  , unsafeNew
+  , newCompleted
```
:+1:

mheinzel commented on src/Database/LSMTree/Internal/MergeSchedule.hs:
```diff
@@ -872,7 +873,7 @@ mergeRuns resolve hfs hbio caching alloc runPaths mergeLevel runs = do
 supplyCredits ::
      (MonadSTM m, MonadST m, MonadMVar m, MonadMask m)
   => TableConfig
-  -> Credit
+  -> UnscaledCredits
   -> Levels m h
   -> m ()
 supplyCredits conf c levels =
```
> calling it `UnscaledCredits` here and `Credits` in the `MergingRun` module feels like it is the wrong way around

I would argue calling it `ScaledCredits` in `MergingRun` feels like the wrong way around since the merging run itself doesn't have anything to do with scaling. For example when we use it inside the `MergingTree`, there is no scaling at all.

> maybe we should have separate `Credits` types with the same name in each module, and rely on qualified imports

That has has potential to be a little confusing, but it might be the best solution.

mheinzel replied:
(done)

mheinzel commented on src/Database/LSMTree/Internal/MergeSchedule.hs:
```diff
@@ -92,11 +91,12 @@ data MergeTrace =
       MergePolicyForLevel
       Merge.Level
-  | TraceCompletedMerge
+  | -- | This is traced at the latest point the merge could complete. To trace
+    --   at the exact point it really completed, we'd need to do it in the
+    --   @MergingRun@ module.
+    TraceExpectCompletedMerge
```
Ah, is that handled correctly by Haddock now? I vague remember it being an issue because the comment is not next to the constructor is is supposed to describe (the `|` is in the way).

mheinzel replied:
Seems like this got fixed at some point, verified by adding `{-# OPTIONS_GHC -Winvalid-haddock -haddock #-}`.

dcoutts commented on src/Database/LSMTree/Internal/MergeSchedule.hs:
```diff
@@ -723,13 +724,19 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
-        mr <- case mergeMaybe of
-          Nothing -> error "newMerge: merges can not be empty"
-          Just m  -> allocateTemp reg (MR.new mergePolicy rs m) releaseRef
+        -- TODO: There currently is a resource management bug that happens if an
+        -- exception occurs after calling MR.new. In this case, all changes roll
+        -- back, so some of the runs in rs will live in the Levels structure at
+        -- their original places again. However, we passed their references to
+        -- the MergingRun, which gets aborted, releasing the run references.
+        -- Instead of passing the original references into newMerge, we have to
+        -- duplicate the ones that previously existed in the level and then
+        -- freeTemp the original ones. This way, on the happy path the result is
+        -- the same, but if an exception occurs, the original references do not
+        -- get released.
```
I agree, we should do exactly what you say here. And that's consistent with what Joris says above: don't adopt references, create new ones and the caller should remove the old ones. I think generally we should try to follow the pattern that the component that allocated the ref then owns it and so also releases it, and we don't transfer ownership.

mheinzel replied:
Done.

mheinzel commented on src/Database/LSMTree/Internal/MergingRun.hs:
```diff
@@ -0,0 +1,496 @@
+  rnf MergeKnownCompleted = ()
+  rnf MergeMaybeCompleted = ()
+
+{-# SPECIALISE new :: HasFS IO h -> HasBlockIO IO h -> ResolveSerialisedValue -> Run.RunDataCaching -> RunBloomFilterAlloc -> Merge.Level -> MergePolicyForLevel -> RunFsPaths -> V.Vector (Ref (Run IO h)) -> IO (Ref (MergingRun IO h)) #-}
```
Not a strong opinion, but I think I prefer the single line versions. There's already the regular type signature, so the specialised version doesn't really contain information I care about (except for being able to verify it's the right specialisation). So it taking up 5+ lines of space just leads to less of the relevant code fitting on my screen.

Ideally we'd have `{-# SPECIALISE new @IO #-}`, but that's only going to be possible in future GHC versions.

jorisdral replied:
The line wrapping was changed in https://github.com/IntersectMBO/lsm-tree/pull/396. I don't have a strong opinion, but you'd be deviating from the status quo

mheinzel replied:
Depends on what is meant by "status quo". Some modules use a single line, some multiple, some are mixed. After moving the code it was about half-half in the new module, so I at least wanted consistency within the module and chose the version I prefer. I can change it, but will probably run into a few conflicts (unless I first squash all commits).

dcoutts reviewed (commented):
Looking good.

Some of the tricky bits here are the resource management. I think it's worth reiterating something I mentioned in a comment below:

I think generally we should try to follow the pattern that the component that allocated the ref then owns it and so also releases it, and we don't transfer ownership. Instead of transferring ownership, duplicate and the new component owns the new reference(s).

mheinzel commented on src/Database/LSMTree/Internal/MergingRun.hs:
```diff
@@ -0,0 +1,496 @@
+        -- If we previously performed too many merge steps, then we perform
+        -- fewer now.
+        let stepsToDo = max 0 (spentCredits + c - totalSteps)
+        -- Merge.steps guarantees that @stepsDone >= stepsToDo@ /unless/ the
+        -- merge was just now finished.
+        (stepsDone, stepResult) <- Merge.steps m stepsToDo
+        assert (case stepResult of
+                  MergeInProgress -> stepsDone >= stepsToDo
+                  MergeDone       -> True
+                ) $ pure ()
+
+        -- This should be the only point at which we write to these variables.
+        --
+        -- It is guaranteed that @totalSteps' >= spentCredits'@ /unless/ the
+        -- merge was just now finished.
+        let totalSteps' = totalSteps + stepsDone
+        let spentCredits' = spentCredits + c
+        -- It is guaranteed that
+        -- @readPrimVar totalStepsVar >= readPrimVar spentCreditsVar@,
+        -- /unless/ the merge was just now finished.
```
It was already used in some parts and I do want some kind of quotation of code for readability. I can use backticks, but now we have two different notations between Haddock and non-Haddock comments. So I thought `@` would be the best solution overall. What do you think?

mheinzel commented on src/Database/LSMTree/Internal/Snapshot.hs:
```diff
@@ -321,56 +320,45 @@ fromSnapLevels reg hfs hbio conf@TableConfig{..} uc resolve dir (SnapLevels leve
+
+              SnapOngoingMerge runs spentCredits lvl -> do
+                rn <- uniqueToRunNumber <$> incrUniqCounter uc
+                mr <- allocateTemp reg
+                  (MR.new hfs hbio resolve caching alloc lvl mpfl (mkPath rn) runs)
+                  releaseRef
```
True. It's slightly different, because here we don't roll back to a state where these runs existed, but we still run into a double free because they were created using `allocateTemp`, which already wants to clean them up.

mheinzel commented on test/Test/Database/LSMTree/Internal/Snapshot/Codec.hs:
```diff
@@ -287,12 +287,12 @@ deriving newtype instance Arbitrary RunNumber
   arbitrary = oneof [
         SnapMergingRun <$> arbitrary <*> arbitrary <*> arbitrary
-                       <*> arbitrary <*> arbitrary <*> arbitrary
+                       <*> arbitrary <*> arbitrary
       , SnapSingleRun <$> arbitrary
       ]
-  shrink (SnapMergingRun a b c d e f) =
-      [ SnapMergingRun a' b' c' d' e' f'
-      | (a', b', c', d', e', f') <- shrink (a, b, c, d, e, f) ]
+  shrink (SnapMergingRun a b c d e) =
+      [ SnapMergingRun a' b' c' d' e'
+      | (a', b', c', d', e') <- shrink (a, b, c, d, e) ]
   shrink (SnapSingleRun a)  = SnapSingleRun <$> shrink a
```
Good catch, same for `Encode`/`DecodeVersioned`.

mheinzel commented on src/Database/LSMTree/Internal/MergingRun.hs:
```diff
@@ -0,0 +1,496 @@
+  rnf MergeMaybeCompleted = ()
+
+{-# SPECIALISE new :: HasFS IO h -> HasBlockIO IO h -> ResolveSerialisedValue -> Run.RunDataCaching -> RunBloomFilterAlloc -> Merge.Level -> MergePolicyForLevel -> RunFsPaths -> V.Vector (Ref (Run IO h)) -> IO (Ref (MergingRun IO h)) #-}
+-- | Create a new merging run, returning a reference to it that must ultimately
+-- be released via 'releaseRef' or by calling 'expectCompleted'.
```
Done in https://github.com/IntersectMBO/lsm-tree/pull/498/commits/90cac301f287c6b39ffdf369bb205eec9cc1114c

mheinzel commented on src/Database/LSMTree/Internal/MergingRun.hs:
```diff
@@ -0,0 +1,448 @@
+  There is an important invariant that we maintain, even in the presence of
+  async exceptions: @merge steps actually performed >= recorded merge steps
+  performed >= recorded spent credits@. TODO: and this makes it correct (?).
+-}
```
Fixed.

mheinzel commented:
I rebased and addressed all points except for the reference handling issues in `MergeSchedule.addRunToLevels` and `MergeSchedule.fromSnapLevels`. I'll continue with those, but I think they can/should go into a separate PR, since the issues existed before and this PR already is large enough. So this PR might be merge-able already, allowing to avoid merge conflicts.

mheinzel commented on src/Database/LSMTree/Internal/Snapshot.hs:
```diff
@@ -321,56 +320,45 @@ fromSnapLevels reg hfs hbio conf@TableConfig{..} uc resolve dir (SnapLevels leve
+                -- completes in supplyMergeCredits, then the resulting run is
+                -- not tracked in the registry, and closing the input runs is
+                -- also not tracked in the registry. Note, however, that this
+                -- bit of code is likely to change in #392.
```
That TODO got fixed in https://github.com/IntersectMBO/lsm-tree/pull/498/commits/4ca09fd0543e9e7f0740b95feea7cd80d46fa7df, so I now also removed the comment there.

jorisdral commented on src/Database/LSMTree/Internal/MergeSchedule.hs:
```diff
@@ -114,6 +94,7 @@ data MergeTrace =
   | TraceCompletedMerge
       NumEntries -- ^ Size of output run
       RunNumber
```
I just noticed we're only tracing this message in the `OneShot` case. Maybe we should trace it in the `Incremental` case as well?

mheinzel replied:
I also noticed it. It's part of a bigger question about tracing and would require a more invasive change, like moving at least part of the `MergeTrace` type into the `MergingRun` module. I can add a TODO at least.

jorisdral commented on src/Database/LSMTree/Internal/MergeSchedule.hs:
```diff
@@ -839,21 +693,25 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
       pure r
 
-    -- TODO: refactor, pull to top level?
+    -- Takes ownership of the runs passed.
```
Might as well remove this comment, since it's not going to be true once we fix the TODO below

mheinzel replied:
IMO it makes the commit that fixes it clearer (we see in the diff what the old behaviour was). If it ends up squashed it doesn't matter either way.

mheinzel replied:
Ah, looking at this again, `newMerge` here actually consumes the references. Otherwise we'd have to do the `releaseTemp` at each call site instead, 6 times. Either way, it's up to date with the implementation.

jorisdral reviewed (commented):
> I rebased and addressed all points except for the reference handling issues in `MergeSchedule.addRunToLevels` and `MergeSchedule.fromSnapLevels`. I'll continue with those, but I think they can/should go into a separate PR, since the issues existed before and this PR already is large enough. So this PR might be merge-able already, allowing to avoid merge conflicts.

You fixed the issue in `fromSnapLevels`, did you not?

mheinzel commented:
I pushed a fix for the reference handling issues.

> You fixed the issue in `fromSnapLevels`, did you not?

Now I did, but before my commit from today, an exception after creating a MergingRun would still have lead to a double free.

mheinzel commented on src/Database/LSMTree/Internal/MergeSchedule.hs:
```diff
@@ -843,18 +696,20 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
       pure r
 
-    -- TODO: refactor, pull to top level?
+    -- | Takes ownership of the runs passed.
```
Done.

mheinzel commented on src/Database/LSMTree/Internal/MergingRun.hs:
```diff
@@ -0,0 +1,496 @@
+-- | Create a new merging run, returning a reference to it that must ultimately
+-- be released via 'releaseRef' or by calling 'expectCompleted'.
+--
+-- Takes over ownership of the references to the runs passed.
```
That's part of the fix, done.

jorisdral commented on src/Database/LSMTree/Internal/MergeSchedule.hs:
```diff
@@ -717,7 +725,10 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
             TraceNewMergeSingleRun (Run.size r)
                                    (Run.runFsPathsNumber r)
-          pure (Single r)
+          -- We create a fresh reference and release the original one.
+          -- This will also make it easier to trace back where it was allocated.
+          freeTemp reg (releaseRef r)
+          Single <$> allocateTemp reg (dupRef r) releaseRef
```
:+1:

jorisdral commented on src/Database/LSMTree/Internal/MergingRun.hs:
```diff
@@ -130,28 +130,36 @@ new ::
-      OngoingMerge runs spentCreditsVar merge
+new hfs hbio resolve caching alloc mergeLevel mergePolicy runPaths inputRuns = do
+    -- If creating the Merge fails, we must release the reference again.
+    bracketOnError (V.mapM dupRef inputRuns) (V.mapM_ releaseRef) $ \runs -> do
```
I mentioned it in https://github.com/IntersectMBO/lsm-tree/pull/498#discussion_r1886597205 too, but this is not safe because each of the `dupRef` operations and that would lead to forgotten references. Similary for `releaseRef`, if one `releaseRef` fails, then it won't try to release the other references

mheinzel replied:
Let's continue the conversation there.

jorisdral commented on src/Database/LSMTree/Internal/MergeSchedule.hs:
```diff
@@ -726,16 +737,9 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
+        V.forM_ rs $ \r -> freeTemp reg (releaseRef r)
         mr <- allocateTemp reg
           (MR.new hfs hbio resolve caching alloc mergeLevel mergePolicy runPaths rs)
           releaseRef
```
It's not important for correctness, but it might be nicer for uninitiated readers if these releases happen after the new merging run is created -- that's closer to how the code would look if we didn't have a temp registry: can't release resources before we use them

mheinzel replied:
done

jorisdral commented on src/Database/LSMTree/Internal/MergeSchedule.hs:
```diff
@@ -705,7 +713,7 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
       pure r
 
-    -- Takes ownership of the runs passed.
+    -- Consumes/releases the runs.
```
I'd just say `release`

jorisdral commented on src/Database/LSMTree/Internal/MergeSchedule.hs:
```diff
@@ -693,6 +700,7 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
             ir' <- newMerge LevelLevelling Merge.LastLevel ln (rs' `V.snoc` r)
             pure $! Level ir' V.empty `V.cons` V.empty
 
+    -- Consumes/releases the incoming run.
```
Same here

jorisdral commented on src/Database/LSMTree/Internal/MergeSchedule.hs:
```diff
@@ -645,6 +645,13 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
     -- NOTE: @go@ is based on the @increment@ function from the
     -- @ScheduledMerges@ prototype.
+    --
+    -- Consumes the runs.
```
Same here

mheinzel commented on src/Database/LSMTree/Internal/MergingRun.hs:
```diff
@@ -162,20 +161,17 @@ unsafeNew mergePolicy mergeNumRuns mergeNumEntries knownCompleted state = do
-  -> Ref (MergingRun m h)
+     (PrimMonad m, MonadMVar m)
+  => Ref (MergingRun m h)
   -> m (V.Vector (Ref (Run m h)))
```
True, it's better to use the TempRegistry here (and also for `new`/`newCompleted`). Should I create a new TempRegistry locally here? It seems nicer than passing it across the module boundary. More modular, we don't have to distinguish in the type signature which functions internally allocate multiple resources and which don't.

mheinzel replied:
Yeah, shouldn't be too bad to nest it. Did that now.

jorisdral reviewed (approved):
LGTM! Let's clean up the commit history bit, and then I'm okay with merging this PR

mheinzel commented:
Squashed a few commits, I think it makes sense to separate them as they are now.
